homeassistant/components/shelly/switch.py

```diff
@@ -0,0 +1,167 @@
+"""Representation of Shelly switches."""
+from homeassistant.components.switch import SwitchEntity
+from homeassistant.const import CONF_DEVICE_ID
+from homeassistant.core import callback
+
+from .const import (
+    COMMAND_SUFFIX,
+    CONF_MODEL,
+    CONF_TOPIC,
+    DOMAIN,
+    MODEL_SWITCHES,
+    MODEL_TITLE,
+    MODELS,
+)
+
+ON = "on"
+OFF = "off"
+TOGGLE = "toggle"
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up Shelly switch(es) from config entry."""
+
+    device_id = config_entry.data[CONF_DEVICE_ID]
+    topic = config_entry.data[CONF_TOPIC]
+    model = config_entry.data[CONF_MODEL]
+    title = MODELS[model][MODEL_TITLE]
+    switches_count = MODELS[model][MODEL_SWITCHES]
+    switches = []
+    for i in range(0, switches_count):
+        switch_topic = topic + f"relay/{i}"
+        if i == 0:
+            unique_id = device_id
+            via_device = None
+        else:
+            unique_id = f"{device_id}_{i+1}"
+            via_device = device_id
+        switches.append(
+            ShellySwitch(
+                hass.components.mqtt, unique_id, switch_topic, title, via_device
+            )
+        )
+
+    async_add_entities(switches)
+
+
+class ShellySwitch(SwitchEntity):
+    """Representation of a Shelly switch."""
+
+    def __init__(self, mqtt, unique_id, topic, title, via_device):
+        """Initialize the MQTT switch."""
+        self._mqtt = mqtt
+        self._unique_id = unique_id
+        self._via_device = via_device
+        self._topic = topic
+        self._title = title
+        self._power = None
+        self._energy = None
+        self._unsubscribe = None
+        self._state = None
+
+    async def async_added_to_hass(self):
+        """Subscribe to MQTT events."""
+        await super().async_added_to_hass()
+        await self._subscribe_topics()
+
+    async def _subscribe_topics(self):
+        """(Re)Subscribe to topics."""
+
+        @callback
+        def state_message_received(msg):
+            """Handle new MQTT state messages."""
+            if msg.topic == self._topic:
+                payload = msg.payload
+                if payload == ON:
+                    self._state = True
+                elif payload == OFF:
+                    self._state = False
+                # how do we handle the "overpower" state?
+            elif msg.topic == self._topic + "/power":
```

Comment:
This topic isn't correct for Shelly 2. Shelly 2 uses `shellies/shellyswitch-<ID>/relay/power'. This one device uses topics without a relay number.

Reply:
Nice catch!
It seems that with Shelly 2, power and energy are for the device, not for a specific relay, so in this case a separate sensor seems like a better approach. It will be inconsistent with e.g. Shelly 2.5, but the underlying devices are inconsistent...